include/boost/spirit/home/x3/binary/binary.hpp

```diff
@@ -0,0 +1,209 @@
+#include <boost/endian/arithmetic.hpp>
+#include <boost/fusion/include/at.hpp>
+#include <boost/mpl/or.hpp>
+#include <boost/type_traits/is_integral.hpp>
+#include <boost/type_traits/is_enum.hpp>
+#include <boost/type_traits/is_floating_point.hpp>
+#include <boost/config.hpp>
+
+namespace boost { namespace spirit { namespace x3
+{
+#if 0
+    ///////////////////////////////////////////////////////////////////////
+    template <typename Encoding, typename Attribute>
+    struct get_info<literal_char<Encoding, Attribute>>
+    {
+        typedef std::string result_type;
+        std::string operator()(literal_char<Encoding, Attribute> const& p) const
+        {
+            return '\'' + to_utf8(Encoding::toucs4(p.ch)) + '\'';
+        }
+    };
+    template <BOOST_SCOPED_ENUM(boost::endian::endianness) bits>
+        struct what;
+
+        template <>
+        struct what<boost::endian::endianness::native>
+        {
+            static std::string is()
+            {
+                return "native-endian binary";
+            }
+        };
+
+        template <>
+        struct what<boost::endian::endianness::little>
+        {
+            static char const* is()
+            {
+                return "little-endian binary";
+            }
+        };
+
+        template <>
+        struct what<boost::endian::endianness::big>
+        {
+            static char const* is()
+            {
+                return "big-endian binary";
+            }
+        };
+    }
+#endif
+    ///////////////////////////////////////////////////////////////////////////
+    template <typename V, typename T
+      , BOOST_SCOPED_ENUM(boost::endian::order) endian, std::size_t bits>
+    struct binary_lit_parser
+      : parser<binary_lit_parser<V, T, endian, bits> >
+    {
+        static bool const has_attribute = false;
+
+        binary_lit_parser(V n_)
+          : n(n_) {}
+
+        template <typename Iterator, typename Context, typename Attribute>
+        bool parse(Iterator& first, Iterator const& last
+          , Context& context, unused_type, Attribute& attr_param) const
+        {
+            x3::skip_over(first, last, context);
+
+            // Even if the endian types are not pod's (at least not in the
+            // definition of C++03) it seems to be safe to assume they are
+            // (but in C++0x the endian types _are_ PODs).
+            // This allows us to treat them as a sequence of consecutive bytes.
+            boost::endian::endian_arithmetic<endian, T, bits> attr_;
+
+#if defined(BOOST_MSVC)
+// warning C4244: 'argument' : conversion from 'const int' to 'foo', possible loss of data
+#pragma warning(push)
+#pragma warning(disable: 4244)
+#endif
+            attr_ = n;
+#if defined(BOOST_MSVC)
+#pragma warning(pop)
+#endif
+
+            unsigned char* bytes = reinterpret_cast<unsigned char*>(&attr_);
+
+            Iterator it = first;
+            for (unsigned int i = 0; i < sizeof(attr_); ++i)
```

Comment:
Should this be done in the constructor? The user would pay a small penalty if the parse function was never called, but if it were called more than once, it would be a small benefit.

Similarly, the move_to function below does any endianness conversion lazily. Its possible to do the conversion immediately in the constructor too. However, a native type could be storing a value in the wrong endianness (which goes against the purpose of the `boost::endian` library).

If the v2 version does it this way, perhaps it should duplicated.